frame/staking/src/scheduler.rs

```diff
@@ -0,0 +1,294 @@
+// TODO: move this to frame/support/scheduler.rs
+
+use sp_std::prelude::*;
+use frame_support::weights::Weight;
+use codec::{Encode, Decode};
+use sp_runtime::traits::Zero;
+use sp_runtime::RuntimeDebug;
+
+/// Definition of task within the runtime.
+///
+/// Each task is represented by something that implements this trait, and that certain something
+/// needs be placed in storage for future scheduling. Therefore, it is required for the task type to
+/// be `Encode + Decode + Default`.
+// TODO: document that you might need to implement some stuff manually here, if you want to be
+// generic over `T`.
+pub trait RuntimeTask: Sized + Clone + Default + Encode + Decode {
+	/// Execute the task while consuming self. The task may at most consume `max_weight` units of
+	/// weight, but could potentially be less.
+	///
+	/// A tuple is returned, where the items are as follows:
+	///   1. Option<Self>, where `None` means that this task is now complete (and shall not be kept
+	///      in storage anymore), and `Some(_)` indicating that this task is not yet complete.
+	///   2. The actual amount of weight that was consumed. Must always be less than `max_weight`.
+	///      parameter.
+	///
+	/// It is critically important for a task to only return a positive consumed weight if it
+	/// _actually did something_. If a positive weight is returned, then a scheduler could interpret
+	/// this as a task that could use another execution slot, and continue the execution potentially
+	/// for numerous iterations.
+	fn execute(self, max_weight: Weight) -> (Option<Self>, Weight);
+}
+
+/// A wrapper for grouping storage-stored tasks together and executing them with a max weight.
+///
+/// A user may create an instance of this struct, add tasks to it (each of which must implement
+///	[`RuntimeTask`]), place it in storage, and execute tasks via the [`execute`] method.
+///
+/// This scheduler is slightly opinionated, and cane be potentially re-implemented with different
+/// strategies. Namely, the assumption is that tasks are heterogenous, meaning that each might
+/// require different weight or be in different stage of execution. See [`execute`] for more info.
+#[derive(Encode, Decode, Default, RuntimeDebug)]
+pub struct RuntimeTaskScheduler<Task: RuntimeTask> {
+	/// The queue of tasks.
+	pub(crate) tasks: Vec<Task>
+}
+
+impl<Task: RuntimeTask> RuntimeTaskScheduler<Task> {
+	/// Create a new (empty) instance of [`RuntimeTaskScheduler`].
+	pub fn new() -> Self {
+		Self { tasks: vec![] }
+	}
+
+	/// Add a new task to the internal queue.
+	pub fn add_task(&mut self, task: Task) {
+		self.tasks.push(task)
```

Comment:
We would need to cap the size of the queue because the weight to decode the `Vec` is linear to the size of the queue. Also we should make use of `append` here to have `O(1)` costs when pusing items to storage.

Reply:
Yeah so in the new API, ideally we need both `append` and `decode_len` in `StoredExecutor`.

Although, this situation exists because I want to abstract the task management from the pallet. As you see now, as a pallet you just accept a `type E: StoredExecutor`, put it in storage and mutate it whenever you want to `.execute`. This is a okay-ish abstraction (compared to you manually being in charge of putting a `Vec<Task>` in storage), but we kinda miss the benefit of `decode_len` and `append`. 

but I am sure we can build these abstractions, if the whole idea of having `StoredExecutor` seems like a good way to go.